api/src/routes/settings.ts

```diff
@@ -163,6 +166,99 @@ export const settingRoutes: FastifyPluginCallbackTypebox = (
     }
   );
 
+  fastify.put(
+    '/update-my-username',
+    {
+      schema: {
+        body: Type.Object({
+          username: Type.String({ minLength: 3, maxLength: 20 })
+        }),
+        response: {
+          200: Type.Object({
+            message: Type.String(),
+            type: Type.Union([Type.Literal('success'), Type.Literal('info')]),
+            username: Type.String()
+          }),
+          500: Type.Object({
+            message: Type.String(),
+            type: Type.Literal('danger')
+          })
+        }
+      }
+    },
+    async (req, reply) => {
+      try {
+        const user = await fastify.prisma.user.findFirst({
+          where: { id: req.session.user.id }
+        });
+
+        const newUsername = req.body.username.toLowerCase();
+        const oldUsername = user?.username.toLowerCase();
+
+        const newUsernameDisplay = req.body.username.trim();
+        const oldUsernameDisplay = user?.usernameDisplay?.trim();
+
+        const alreadyUsername =
+          newUsername === oldUsername &&
+          newUsernameDisplay === oldUsernameDisplay;
+
+        if (alreadyUsername && oldUsernameDisplay) {
+          return {
+            message: 'flash.username-used',
+            type: 'info',
+            username: newUsernameDisplay
+          } as const;
+        }
+
+        const validation = isValidUsername(newUsername);
+
+        if (!validation.valid) {
+          return {
+            message: `Username ${newUsername} ${validation.error}`,
+            type: 'info',
+            username: newUsernameDisplay
+          } as const;
+        }
+
+        const hasProfanity = new badWordsFilter().isProfane(newUsername);
+        const preserved = blocklistedUsernames.includes(newUsername);
+
+        // Checks for both username and usernameDisplay because users
+        // can have the same username but with differnt casing
+
+        const exists = await fastify.prisma.user.findFirst({
+          where: { username: newUsername, usernameDisplay: newUsernameDisplay }
+        });
+
+        if (exists || hasProfanity || preserved) {
```

Comment:
Just out of curiosity, do we not want to display a separate message for profanity?

Reply:
That is not what we currently do.